SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialFunctionalTest.cpp

```diff
@@ -229,6 +229,39 @@ void ASpatialFunctionalTest::RegisterFlowController(ASpatialFunctionalTestFlowCo
 		//FlowControllers invoke this on each worker's local context when checkout and ready, we only want to act in the authority
 		return;
 	}
+
+	if (FlowController->ControllerType == ESpatialFunctionalTestFlowControllerType::Client)
+	{
+		// Because of the way interest borders between server workers are setup, you can have clients being spawned outside the Test's
+		// server worker, which means that it will have an id given locally by the server worker where the client spawned. So when registering,
+		// the Test server worker will ensure that the clients all have unique ids. Note that server workers don't have this issue
+		// because their ids are tied to their virtual worker id, hence guaranteed to be unique.
+		uint8 DesiredClientId = FlowController->ControllerInstanceId;
+		while (true)
+		{
+			bool bFoundConflictingId = false;
+			for (auto* AuxController : FlowControllers)
+			{
+				if (AuxController->ControllerType == ESpatialFunctionalTestFlowControllerType::Client && AuxController->ControllerInstanceId == DesiredClientId)
+				{
+					bFoundConflictingId = true;
+					DesiredClientId += 1; // there's a collision, let's try the next one
+					checkf(DesiredClientId != 0, TEXT("This situation should only happen if you have more than 255 players, so you'll probably already be in trouble given that we don't support that"));
+					break;
+
+				}
+			}
+			if (!bFoundConflictingId)
+			{
+				if (FlowController->ControllerInstanceId != DesiredClientId)
+				{
+					FlowController->ControllerInstanceId = DesiredClientId;
```

Comment:
Do controllers need an up to date ID in the authoritative execution context when asked to execute a step? eg. if controller is asked to run step, will it do a local check that it should run the step?

I don't think that's the case directly, but maybe in some of the tests that "do something for client X"?

If that's the case, we shouldn't set the value locally, and should rely on it being set by the crossserver RPC, and replicating back to us, as a Test::IsReady() condition

Reply:
It is set directly so that if the test starts, it will already have the proper id. Otherwise we need to have a way to verify that this has propagated, because when you start a step it checks against the id.

If I don't set it locally here, then the only other way I can see this working is if by default they have an invalid id (e.g. 0), then they get always assigned by the authority of the test (like requested above), and the test only starts when all of them don't have an invalid id anymore in the authority.

Reply:
I'll rework it to be this way.